src/features/environmentDetails/components/Specification/SpecificationEdit.tsx

```diff
@@ -116,7 +131,7 @@ export const SpecificationEdit = ({
       dispatch(updatePackages(code.dependencies));
       dispatch(updateChannels(code.channels));
     } else {
-      setCode({ dependencies: requestedPackages, channels });
+      setCode({ dependencies: requestedPackages, variables, channels });
```

Comment:
Why is there no dispatch line for variables in the if-branch just above?

Reply:
Because you can only set variables by editing the YAML specification. There's no dedicated UI element for variables, which is what the UI block above is for. See the attached screenshots:

<img width="894" alt="Screen Shot 2024-01-31 at 18 17 49" src="https://github.com/conda-incubator/conda-store-ui/assets/10469885/ea8e78c2-42d0-4d06-942b-c19bf19c9c24">
vs.
<img width="901" alt="Screen Shot 2024-01-31 at 18 17 40" src="https://github.com/conda-incubator/conda-store-ui/assets/10469885/051f11e6-2ce1-4630-be6e-d1b263e3a9a5">

Reply:
Yes, I thought that might be the case. However, the thing that's confusing is that `dispatch` is about updating the global store, which isn't UI.

So even though there isn't a dedicated GUI element to surface the state changes made to the `variables` key in the YAML editor, you should still update the app state (unless I'm still missing something here). Why? Imagine that somebody else comes into the codebase whose task is to add some component that displays the value of `variables`. They understand that `variables` can be updated somehow, but how it's updated is not really their concern. They're just tasked with writing the component that displays the value of `variables`. So they would hook into the global store, read the `variables` key, and display the values. But if the code for the YAML editor never updates the global store (via `dispatch`), then our programmer has just coded up a completely useless component that never displays the value of `variables`.

As an aside, it looks to me that the calls to `dispatch` are in an odd place. I wonder why `dispatch` isn't called on editor update.

Reply:
The actionable part of my reply: I think the GUI parts can be implemented separately. This was intended as a quick fix for a problem that some users were having. More details below, feel free to skip.

----

#### How state updates work

The way this currently works:
- if you make a change to the raw YAML, the variables are added to the app state
- if you make a change via the GUI, the current variables state is used (could be the initial empty state, could be the one set via raw YAML before).

#### Why updating state in raw YAML seems fine

To me, this seems better than updating variables' state on GUI updates (which would need to remove previously set values since we have no GUI elements for variables at the moment). Using GUI is convenient for adding packages. Setting variables is not something that should happen often. I also wanted to make as few changes as possible to keep this well-scoped, so this fits nicely into existing logic.

#### Why variables are needed

The variables support was supposed to be a quick fix to unblock Nebari users. The initial version of this PR was ready right after it was implemented on the server. It's just that we had no one to review the UI part for a bit.

On Nebari, users usually spin up CPU-only instances, but they often want to build environments with GPU support anyway, in case that's needed later. Also, conda-store-server currently supports setting only one particular variable (`CONDA_OVERRIDE_CUDA`) and ignores everything else.

#### Process for GUI changes

When GUI changes are required, our process has been to get a design from our designer first, to ensure the best UX. Then someone implements that on the UI side. That's why I only implemented the non-GUI parts of this.

#### Process for avoiding errors

Regarding potential issues if GUI were to be implemented separately: we have a process in place where each PR requires testing and is reviewed by someone on the team with relevant experience.